caml_z.c

```diff
@@ -801,6 +852,31 @@ CAMLprim value ml_z_to_int64(value v)
   return caml_copy_int64(x);
 }
 
+CAMLprim value ml_z_to_int64_unsigned(value v)
+{
+  uint64_t r;
+  Z_DECL(v);
+  Z_MARK_OP;
+  Z_CHECK(v);
+  if (Is_long(v)) {
+    intnat x = Long_val(v);
+    if (x < 0) ml_z_raise_overflow();
+    return caml_copy_int64(x);
+  }
+  Z_MARK_SLOW;
+  Z_ARG(v);
+  if (sign_v) ml_z_raise_overflow();
+  switch (size_v) {
+  case 0: r = 0; break;
+  case 1: r = ptr_v[0]; break;
+#ifndef ARCH_SIXTYFOUR
+  case 2: r = ptr_v[0] | (ptr_v[1] << 32); break;
```

Comment:
This is shifting a 32-bit unsigned type by 32: undefined behavior.
```suggestion
  case 2: r = ptr_v[0] | ((uint64_t) ptr_v[1] << 32); break;
```